src/lib-imap/imap-capability-list.c

```diff
@@ -0,0 +1,127 @@
+#include "imap-capability-list.h"
+
+/* initialize the imap capability list */
+struct imap_capability_list *
+imap_capability_list_create(const char *initial_entries)
+{
+	/* initialize a capability list object */
+	struct imap_capability_list *
+		capability_list = i_malloc(sizeof(struct imap_capability_list));
+	capability_list->refcount = 1;
+	/* initialize the capability list memory pool */
+	capability_list->pool = pool_alloconly_create(MEMPOOL_GROWING"Capability List", 2048);
+	/* initialize the capability list capability array */
+	p_array_init(&(capability_list->imap_capabilities),
+			capability_list->pool,
+			0);
+
+	/* make sure our initial entries are appended */
+	if (initial_entries) {
+		imap_capability_list_append_string(capability_list, initial_entries);
+	}
+
+	return capability_list;
+}
+
+/* increment reference to an imap capability list */
+void imap_capability_list_ref(struct imap_capability_list *capability_list)
+{
+	i_assert(capability_list->refcount > 0);
+
+	capability_list->refcount++;
+}
+
+/* decrement reference to an imap capability list */
+void imap_capability_list_unref(struct imap_capability_list **capability_list)
+{
+	i_assert((*capability_list)->refcount > 0);
+
+	if (--(*capability_list)->refcount > 0)
+		return;
+
+	pool_unref(&(*capability_list)->pool);
+	i_free(*capability_list);
+	*capability_list = NULL;
+}
+
+/* add an entry to the imap capability list */
+void imap_capability_list_add(struct imap_capability_list *capability_list,
+		const char *command, capability_cond_t flags)
+{
+	if (!capability_list || !command) {
+		return;
+	}
+
+	/* is it already added? */
+	if (imap_capability_list_find(capability_list, command) != NULL) {
+		return;
```

Comment:
I wonder if double-adding is a bug or an expected situation, i.e. should it silently return or assert-crash?..